controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,890 @@
+	aksv1 "github.com/rancher/aks-operator/pkg/apis/aks.cattle.io/v1"
+	v10 "github.com/rancher/aks-operator/pkg/generated/controllers/aks.cattle.io/v1"
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v15 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	controllerName           = "aks-controller"
+	controllerRemoveName     = "aks-controller-remove"
+	aksConfigNotCreatedPhase = ""
+	aksConfigActivePhase     = "active"
+	aksConfigUpdatingPhase   = "updating"
+	aksConfigImportingPhase  = "importing"
+	aksClusterConfigKind     = "AKSClusterConfig"
+)
+
+type Handler struct {
+	aksCC           v10.AKSClusterConfigClient
+	aksEnqueueAfter func(namespace, name string, duration time.Duration)
+	aksEnqueue      func(namespace, name string)
+	secrets         wranglerv1.SecretClient
+	secretsCache    wranglerv1.SecretCache
+}
+
+func Register(
+	ctx context.Context,
+	secrets wranglerv1.SecretController,
+	aks v10.AKSClusterConfigController) {
+
+	controller := &Handler{
+		aksCC:           aks,
+		aksEnqueue:      aks.Enqueue,
+		aksEnqueueAfter: aks.EnqueueAfter,
+		secretsCache:    secrets.Cache(),
+		secrets:         secrets,
+	}
+
+	// Register handlers
+	aks.OnChange(ctx, controllerName, controller.recordError(controller.OnAksConfigChanged))
+	aks.OnRemove(ctx, controllerRemoveName, controller.OnAksConfigRemoved)
+}
+
+func (h *Handler) OnAksConfigChanged(key string, config *aksv1.AKSClusterConfig) (*aksv1.AKSClusterConfig, error) {
+	if config == nil || config.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	ctx := context.Background()
+	switch config.Status.Phase {
+	case aksConfigImportingPhase:
+		return h.importCluster(ctx, config)
+	case aksConfigNotCreatedPhase:
+		return h.createCluster(ctx, config)
+	case aksConfigActivePhase, aksConfigUpdatingPhase:
+		return h.checkCluster(ctx, config)
+	default:
+		return config, fmt.Errorf("invalid phase: %v", config.Status.Phase)
+	}
+}
+
+func (h *Handler) OnAksConfigRemoved(key string, config *aksv1.AKSClusterConfig) (*aksv1.AKSClusterConfig, error) {
+	if config.Spec.Imported {
+		logrus.Infof("Cluster [%s] is imported, will not delete AKS cluster", config.Spec.DisplayName)
+		return config, nil
+	}
+
+	ctx := context.Background()
+	logrus.Infof("Removing cluster [%s] for config [%s]", config.Spec.ClusterName, config.Spec.DisplayName)
+
+	azureAuthorizer, err := newClientAuthorizer(h.secretsCache, config.Spec)
+	if err != nil {
+		return config, err
+	}
+	resourceGroupsClient, err := newResourceGroupsClient(azureAuthorizer, config.Spec)
+	if err != nil {
+		return config, err
+	}
+	resourceClusterClient, err := newClustersClient(azureAuthorizer, config.Spec)
+	if err != nil {
+		return config, err
+	}
+
+	if existsCluster(ctx, resourceClusterClient, config.Spec) {
```

Comment:
Should this be right below the imported check so we don't log and initialize all this stuff ?

Reply:
We need `resourceClusterClient` object initiated, so this has to stay as it is